### DISCUSSION TITLE
RFC: useCollapsible hook

### DISCUSSION THREAD

**Discussion by matheusps** [vtex/shoreline#438] [Technical RFCs]
# RFC useCollapsible hook

## Why

The current `collapsible` component does not match the design guidelines. The `card` component can be composed/enhanced to accomplish collapsible content. This proposal means to deprecate the collapsible in favor of a lower-level approach with hooks. This will enhance the flexibility and reuse of disclosure-collapsible behavior. 

## Pros

- Enhance collapsible behavior flexibility.
- Allow collapsible behavior reuse.
- Less component overhead, the Design System API surface area gets smaller.

## Drawbacks

- It gets more complex to learn and apply because the actual use is not attached to a component.
- Migration is needed, and code mods will not help much (good old documentation, them 😀 ).

### Before

We use the `collapsible` component with a state hook.

```jsx
import { useCollapsibleState, Collapsible, CollapsibleHeader, CollapsibleContent } from '@vtex/admin-ui'

function Example() {
  const state = useCollapsibleState()

  return (
    <Collapsible state={state}>
      <CollapsibleHeader label="your label" />
      <CollapsibleContent> your content </CollapsibleContent>
    </Collapsible>
  )
}
```

### After

Replace the `collpasible` with the `useCollapsible` hook, which returns the props (using the prop-getters pattern). 

```jsx
function Example() {
  const { getToggleProps, getCollapseProps } = useCollapsible()

  return (
    <Card>
      <CardHeader {...getToggleProps()}>
        <CardTitle>Your title</CardTitle>
      </CardHeader>
      <CardContent {...getCollapseProps()}>Your content</CardContent>
    </Card>
  )
}
```

#### Abstracting

The developer can abstract the component into a stateful accordion, or whatever makes sense for the application. 

```jsx
// accordion.tsx
interface Props {
  title: ReactNode
  content: ReactNode
}

export function Accordion (props: Props) {
  const { title, content } = props
  const { getToggleProps, getCollapseProps } = useCollapsible()

  return (
    <Card>
      <CardHeader {...getToggleProps()}>
        <CardTitle>{title}</CardTitle>
      </CardHeader>
      <CardContent {...getCollapseProps()}>{content}</CardContent>
    </Card>
  )
}

// usage.tsx
import { Accordion } from './accordion.tsx'

<Accordion title="Your title" content="your content" />
```

**Comment by mairatma** [comment]
I like the new approach better too.

I'd just like to propose that we mitigate the con about this being more complex to apply by adding some utility props for frequently used patterns. For example, if people frequently use collapsible with Card then we could add a `collapsible` optional prop to it that automatically calls `useCollapsible` and applies it to the right child components automatically. Then people could just do:

```jsx
function Example() {

  return (
    <Card collapsible>
      <CardHeader>
        <CardTitle>Your title</CardTitle>
      </CardHeader>
      <CardContent>Your content</CardContent>
    </Card>
  )
}
```

**Reply by matheusps** [reply]
I think this harms the `Card` separation of concerns since it will always know the collapsible behavior, while on the proposal, it is agnostic to that. This is good for the internal component architecture, it keeps things simple and reduces individual components bundle. 

The way I would go is to create a component, CollapsibleCard. But that's the thing we are trying to avoid here 😅. I think that the main problem here is that this behavior is not a pattern yet. In the future, it may.

**Comment by mairatma** [comment]
And about the other con, related to codemods not helping much in this case. Couldn't we create a codemod that outputs some components that mimic the same UI that Collapsible had before, but using the new hook?

**Reply by matheusps** [reply]
The codemod while dealing with state hooks is pretty tricky. The chance of the parsed code being non-functional is way bigger. Especially because you can pick and change behavior from the hook. For example:

```ts
const state = useCollapsibleState()

// ...

useEffect(() => {
  if(state.open) { /** ... */ } 
}, [state.open])
```
In this example, if you remove de useCollapsibleState - the code stops working.

**Reply by matheusps** [reply]
The movement from a "black-box" model to the compound would work like a champ. 
`<CollapsibleHeader label="label" />` to `<CardHeader><CardTitle>label</CardTitle></CardHeader>`

**Reply by mairatma** [reply]
Got it! I didn't know the usage of this `useCollapsibleState`. This indeed makes things trickier. How will people control the state in the new approach you're suggesting by the way? I'm curious now.

But regardless, if using codemods are complicated here, could we at least use the codemod tool (or another) to detect calls to `useCollapsibleState` and `<Collapsible>` and just print a warning/error message on the terminal to explain that these need to be migrated to the new way, but couldn't be done automatically? Maybe that's what's usually done already, but just giving the idea as that would already help people know what they need to do when we can't fix automatically.

**Reply by matheusps** [reply]
The state will still be managed by a hook, but it will get more responsibility. The get functions replace the logic that would be done by the component previously. So, in other words, o the previous version:

- useCollapsibleState: responsible for handling just state
- Collapsible: receive the state, mixes the state properties with props, and passes the HTML props for an element (internally)

On the new approach:

- useCollapsible: handle the state, and returns the get functions.
- getFunctions: the same as the Collapsible component, but return the props as objects.

**Reply by mairatma** [reply]
Got it! Let me know what you think about the idea of warning via the codemod at least, so people can detect and manually update these breaking changes more easily.

**Reply by matheusps** [reply]
Ooops! About that, I don't know... It sounds strange though. Would it be a codemod that doesn't do any code modifications? Maybe it should be an Eslint rule them. Or, we can transform and warn: "This transformation sometimes is not safe, check and review this file". Because that are cases it works, and cases it doesn't. What do you think? 🤔

**Reply by mairatma** [reply]
Agreed! If we don't have any other codemods for the next release it'd be weird to have one that doesn't do anything. If we already have other codemods running I don't think it's weird though, as it would be an extra output, for example: _these things we managed to fix automatically, but there are some points that you need to manually change here and here_.

But regardless of how we do it, I just mean that we should just try to have some way of indicating these kind of breaking changes that we can't automate instead of letting the developer notice the crash and have to figure out how to fix it themselves.

**Comment by mairatma** [comment]
And finally, do these need to be functions: `getToggleProps` and `getCollapseProps`? Couldn't they just be `toggleProps` and `collapseProps`, for example?

**Reply by matheusps** [reply]
I think `get` is good for the user not confusing a function to an object. In the wild (community), this pattern is generally implemented with `get[component-name-Props]` ([downshift benchmark](https://www.downshift-js.com/use-combobox#props-used-in-examples)). But, if we agree without it is better - all good to me.

**Reply by mairatma** [reply]
Oh I totally agree, I was just suggesting that they could be the plain objects instead of functions. Not sure if making them functions has another purpose though, it just seemed like they could be the objects directly instead of functions that return them.

**Reply by matheusps** [reply]
Functions are quite useful for custom props. It gets clearer when you see the implementation of a getter. The toggle care about the onClick property, maybe you want to use it for something else. A simple `getToggleProps` implementation:

```jsx
const [open, setOpen] = useState(false)

const toggle = () => setOpen(o => !o)

getTogglerProps = (props = {}) => ({
  'aria-controls': 'target',
  'aria-expanded': Boolean(open),
  ...props,
  onClick: callAllHandlers(props.onClick, toggle),
})

// callAllHandlers is just a function that calls all handlers that are present, for example:
const callAllHandlers =
  (...functions) =>
  (...args) =>
    functions.forEach(fn => fn && fn(...args))
```

When you are using the function, it gets easy to pass an onClick:

```jsx
<button {...getToggleProps({ onClick: ()=> console.log('customClick') })}>Toggle</button>
```

If you're using objects, not so much

```jsx
const buttonProps = { ...toggleProps, onClick: () => {
  console.log('customClick')
  toggleProps.onClick(),
}}

<button {...buttonProps}>Toggle</button>
```

And this can also go wrong since the user needs to know which props the `toggleProps.onClick` needs to receive. This can warm the function's integrity.

**Reply by mairatma** [reply]
Got it! I basically wanted to know if these functions would receive optional args or not. If it's useful for them to have args, then they're better off as functions, like you suggested. With no args the functions seemed like they could just be objects to further simplify the usage, but that's not the case.

**Reply by matheusps** [reply]
Yeah, without custom values it would simplify a lot. I would keep it this way because custom onClick is a standard feature.

**Comment by anitavincent** [comment]
@matheusps, regarding the example you mentioned

```
const state = useCollapsibleState()


useEffect(() => {
  if(state.open) { /** ... */ } 
}, [state.open])
```

on the proposed api where would this data (formerly state.open) be? Would a call to the getters be necessary? Or would there be more props on the state beside the getters?

To make it simpler the current hook has this api, are you proposing something like it, or different?

```
export interface CollapsibleState {
  getToggleProps: (config?: GetTogglePropsInput) => GetTogglePropsOutput
  getCollapseProps: (config?: GetCollapsePropsInput) => GetCollapsePropsOutput
  /** Is true if content is currently visible */
  visible: boolean
  /** Toggle collapsible expansion */
  toggle: () => void
}
```

**Reply by matheusps** [reply]
The example continues working since the state is not a part of the getters. Would be something like this:

```jsx
const { getToggleProps, open } = useCollapsible()

useEffect(() => {
  if(open) { /** ... */ } 
}, [open])
 
<button {...getToggleProps()} />
```

Makes sense?

**Comment by lafray** [comment]
I thought the decoupling of the collapsible behavior from the card was very important. I've had components in some projects that needed this behavior but that didn't have the collapsible style (I can bring the examples if you want). But we wanted to use the admin component to have the default opening and closing effect. This ended up causing quite a headache. I only see advantages in separating the two concepts. A api proposta me pareceu super intuitiva e conforme outros padrões usados

**Comment by davicostalf** [comment]
@matheusps Can we close this?
